sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -528,6 +528,40 @@ object SimplifyConditionals extends Rule[LogicalPlan] with PredicateHelper {
 }
 
 
+/**
+ * Push the foldable expression into (if / case) branches.
+ */
+object PushFoldableIntoBranches extends Rule[LogicalPlan] with PredicateHelper {
+  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    case q: LogicalPlan => q transformExpressionsUp {
+      case b @ BinaryExpression(i @ If(_, trueValue, falseValue), right) if i.deterministic &&
+          right.foldable && (trueValue.foldable || falseValue.foldable) =>
```

Comment:
This is good idea when both `trueValue` and `falseValue` are foldable like the example in the description.

But if only `trueValue` is foldable? For example, from `If(_, 'b', col1) = 'a'`, we get `If(_, 'b' = 'a', col1 = 'a')`, are some benefits here too?

Reply:
The differences in generated code:
![image](https://user-images.githubusercontent.com/5399861/102587582-0ce8d100-4147-11eb-9ede-4a5604922946.png)

Reply:
Hmm, I'm not sure if the difference means obvious benefits. Looks like it should be very close but I think I'm okay with it.